### PR TITLE
Rename "reference" nonterminal to "smart_global" in documentation

### PR DESCRIPTION
**Pull request by jfehrle**
I wonder if `smart_qualid` might be better.  Just seeing the name, it gives a better idea of the syntax.  Does it really only refer to global entities?

### CODE REVIEW

File: doc/sphinx/language/extensions/arguments-command.rst
Comment (Zimmi48) [inline doc/sphinx/language/extensions/arguments-command.rst:277]:
```suggestion
   turn to be themselves arguments of a global reference are interpreted
   accordingly to the argument scopes bound to the global reference.
```

**Comment by Zimmi48** [comment]
> I wonder if `smart_qualid` might be better. Just seeing the name, it gives a better idea of the syntax. Does it really only refer to global entities?

That was what I initially thought as well but @herbelin and @JasonGross' comments made clear that the semantics was more important in that case.

**Comment by SkySkimmer** [comment]
Are they really all "smart" uses, meaning if I have a definition `foo` and do `Notation bar := foo.` will they understand that `bar` means `foo`?

**Comment by Zimmi48** [comment]
No, apparently "abbreviations" are not supported where global references are expected.

**Comment by jfehrle** [comment]
Updated.

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
@JasonGross and @herbelin: any comment before I merge this?

**Comment by JasonGross** [comment]
> No, apparently "abbreviations" are not supported where global references are expected.

I am still confused by this.  I thought `smart_global` was for the things passed to blacklists/whitelists in `cbv delta`, which does accept abbreviations to indicate the head symbol.

The only place I can think of where we want a qualid without this behavior is `Print`, but perhaps there are other places?

Separately, there is the question of whether or not to accept `"+"%nat` to refer to `Nat.add`.  My inclination is that it's fine to allow this in all cases?  But maybe we need to do something different for `Locate` (or anything else that needs to know whether it's getting a notation or a qualid)?  (Maybe we also want to disallow this for `Print`?)

I think before we merge this PR we should decide which subset of cases we're eventually going to allow, and what the names for all of them should be.  Maybe we should discuss on the Coq call?

**Comment by jfehrle** [comment]
@Zimmi48 Let me know what changes, if any, are needed as a result of the Coq call discussion.

**Comment by Zimmi48** [comment]
IIUC, the main decision is to keep the name "reference", so we could simply close this PR. But we will have to stop using this nonterminal in the `Locate` and `Print` commands because in such cases they are interpreted differently.

**Comment by jfehrle** [comment]
The uses in `Print` and `Locate` are in the `.mlg` files.  Would be nice to fix there rather than more editing with `doc_grammar`--there's already too much.  But not clear to me that anyone agreed to do such changes.

**Comment by Zimmi48** [comment]
Here we are talking about 8.12 documentation changes, so modifying the mlgs is not an option. But making mlgs closer to what's documented will clearly be a mid-term objective.

**Comment by herbelin** [comment]
As far as I'm concerned, I'm pretty ok with the conclusion.

Assuming that this is adopted, here is a first attempt at a possible four-phase protocol to eventually unify the terminology towards only `qualid` and `reference` across the reference manual, the `mlg`s, the various `EXTEND` mechanisms and the `Tactic Notation` mechanism.

*First phase* (which can be implemented as soon as decided)

- In `g_prim.mlg`:

  - keep only `qualid`, deprecate the `G_prim.reference` synonym, and consequently replace all `G_prim.reference` by `G_prim.qualid` in other mlg files

- In `g_ltac.mlg`:
  - introduces a `ltac_reference` for `Print Ltac` and `Locate Ltac` which is an alias for `qualid`
  - in the Ltac grammar, replaces `reference` by `qualid` (since in an Ltac call, the reference is "allowed" to refers also to a term)

- In `g_class.mlg`:
  - in `Typeclasses Transparent` and `Typeclasses Opaque`, generalize `reference_list` into `smart_global_list`?

- In reference manual:
  - change `Locate`, `Print` and `About` so that they don't take a `reference` but a disjunction of `@qualid` and `@string {? % @scope_key }`
  - in Ltac notation chapter: replace entry `reference` by `qualid`, remove `unfold` in the description of `reference` (since `unfold` takes a `smart_global`) (cf #12698)

- In `stdarg.ml`:
  - add a `wit_qualid` with same semantics as current `wit_reference`; change `wit_reference` so that it becomes an equivalent to `wit_smart_global`.

- In Ltac plugin:
  - do so that when `reference` is used in `TACTIC EXTEND` and cie, as well as in `Tactic Notation`, a warning is issued at compilation time telling to use `qualid` if the intent is to support only a qualified identifier or to keep as it is if the intention is to use a smart global. [How to silence the warning though?]

*Second phase* (8.14?):

- Remove `reference` in `g_prim.mlg` (forcing `GRAMMAR EXTEND` plugins to use either `qualid` or `smart_global`); in `Tactic Notation` or `TACTIC EXTEND`, all of `qualid` and of the two synonyms `reference` and `smart_qualid` remain allowed

*Third phase* (8.15?):

- In `g_prim.mlg`:
  - restore `G_prim.reference` in the sense of `smart_global` and replaces all uses of `smart_global` in `mlg` by `reference`
  - deprecate `smart_global` (forcing plugins to eventually use `reference`)

- In reference manual:
  - in Ltac notation chapter: rename entry `smart_global` to (the new) `reference`

- In Ltac plugin:
  - deprecate `smart_global` in `TACTIC EXTEND` and `Tactic Notation` (forcing code to eventually use `qualid` or `reference`)

*Fourth phase*

- Remove remaining deprecated uses of `smart_global`

PS: This protocol would work also with another choice of (final) names.

Optional: also renounce to `global` in the grammars to the benefit of `qualid`, and deprecate `global` in `EXTEND` and cie, asking to use `reference` or `qualid` instead.

**Comment by jfehrle** [comment]
We really need 3 releases (figure 2+ years from now) to make this change?  I'd be worried that we forget to implement phases 2-4.

IIUC OCaml lets you mark variables as deprecated (see https://caml.inria.fr/pub/docs/manual-ocaml/attributes.html and for OCaml 4.08 (which we don't require yet), https://ocaml.org/releases/4.08.0.html.  Could we use this to make the change more apparent to outside code that refers to the existing symbols?  (Maybe this is what you had in mind?)

Could we create a `GPrim.New_reference` symbol in phase 1 instead of waiting until phase 3?  The `New_` would help remind developers that it's been changed while documenting that the prefix is meant to go away.  We could update the documentation sooner without ever showing `New_reference` in the grammar (but maybe mentioned in in the`Tactic Notation` section).

I expect you've thought about this more deeply than I have.  I'm happy to help with the doc updates once the plan is set.

**Comment by jfehrle** [comment]
PS Once I've finished updating the grammars in the doc, there will be other symbols we may consider renaming to make the doc and code more consistent--though almost all simpler than this case where we're reclaiming `reference` so we can use it to mean something else.

**Comment by herbelin** [comment]
> IIUC OCaml lets you mark variables as deprecated (see https://caml.inria.fr/pub/docs/manual-ocaml/attributes.html and for OCaml 4.08 (which we don't require yet), https://ocaml.org/releases/4.08.0.html. Could we use this to make the change more apparent to outside code that refers to the existing symbols? (Maybe this is what you had in mind?)

For `G_prim.reference`, it is maybe possible to use the ocaml deprecation through `coqpp`. We would need to check the details.

For `wit_reference`, I guess it should be handled in `coqpp`
 
> Could we create a `GPrim.New_reference` symbol in phase 1 instead of waiting until phase 3?

I'm fine with this idea. We could then even use it for another such transition, the one of `ident` to `name` in #11841 (i.e., there, we could provide a temporary `new_ident` for one release long).

PS: I'm a bit worry about library developers who try to provide versions of their library available for different versions of Coq. Shouldn't we eventually provide an ifdef mechanism (e.g. `CoqVersion <= 8.12 [ command ]`, or an attribute, or whatever convenient syntax)?

**Comment by Alizter** [comment]
So what happened with this in the end?

**Comment by Zimmi48** [comment]
The conclusion was to keep `reference` in the documentation and maybe at some point migrate the mlg files so that they use the same name.

**Comment by Zimmi48** [comment]
See also the more detailed plan described in https://github.com/coq/coq/pull/12568#issuecomment-658961186.

**Pull request closed without merging**
